tests/deploy/test_scheduler_agent.py

```diff
@@ -3137,12 +3138,199 @@ async def restore_baseline_state() -> None:
     assert scheduler._state.resource_state[rid5].deployment_result is DeploymentResult.NEW
     assert scheduler._state.resource_state[rid5].blocked is BlockedStatus.NO
 
-    # TODO: add more scenarios
-    # - resource becomes undefined (in first, second, or last version)
-    # - resource becomes defined (in first, second, or last version)
-    # - resource becomes undefined and then defined again and vice versa
-    # - resource is new+undefined / updated+undefined
-    # - verify requires => like resources, should simply be `== all_models[-1].requires`, same for provides
+    await restore_baseline_state()
+
+    # Resource becomes undefined
+    await scheduler._new_version(
+        [
+            model(
+                {
+                    rid1: all_models[-1].resources[rid1],
+                    rid2: all_models[-1].resources[rid2],
+                    rid4: make_resource_minimal(rid4, values={"new": "unknown"}, requires=[]),
+                },
+                undefined={rid4},
+            ),
+            model(
+                {
+                    rid1: make_resource_minimal(rid1, values={"changed": "unknown"}, requires=[]),
+                    rid2: all_models[-1].resources[rid2],
+                    rid4: make_resource_minimal(rid4, values={"new": "unknown"}, requires=[]),
+                },
+                undefined={rid1, rid4},
+            ),
+            model(
+                {
+                    rid1: make_resource_minimal(rid1, values={"changed": "unknown"}, requires=[]),
+                    rid2: make_resource_minimal(rid2, values={"changed": "unknown"}, requires=[]),
+                    rid4: make_resource_minimal(rid4, values={"new": "unknown"}, requires=[]),
+                },
+                undefined={rid1, rid2, rid4},
+            ),
+        ]
+    )
+    assert scheduler._state.resources == all_models[-1].resources
+    for rid in (rid1, rid2, rid4):
+        assert scheduler._state.resource_state[rid].status is ComplianceStatus.UNDEFINED
+        assert (
+            scheduler._state.resource_state[rid].deployment_result is DeploymentResult.DEPLOYED
+            if rid != rid4
+            else DeploymentResult.NEW
+        )
+        assert scheduler._state.resource_state[rid].blocked is BlockedStatus.YES
+
+    await restore_baseline_state()
+
+    # Resource becomes defined
```

Comment:
This one is now more or less the same as the next scenario ("Resource becomes undefined, defined and then undefined again"). I think it would be good to have at least one resource that starts as undefined (i.e. from a previous `_new_version`), to make sure we cover that scenario as well. e.g. through a separate `_new_version()` call to setup the initial scenario, and only then have the multi-version `_new_version()` call.